Biggems' Mythical Bug Bashing 2017 Edition ( exclusive )

runedog48 opened the pull request:
A collection of various bug fixes.

- Virology Isolation Centrifuge's icon will update properly based on its current state ( processing or idle )
- Copied papers will no longer appear blank
- Fixed #13374 ( Taken from Aurora credit to skull132 )
- Fixed #19912
- Fixed #17281
- Fixed #17429
- Fixed #17416
- Fixed #19799
- Fixed #17346
- Fixed #17968 ( multiple stuck objects are condensed into ‘multiple [object name]’
- Fixed #17326
- Fixed #19897 ( bandaid )
- Fixed #17609

Commit:
Multiple Bug fixes
- Virology Isolation Centrifuge's icon will update properly based on its current state ( processing or idle )
- Fixed #13374 ( Taken from Aurora credit to skull132 )
- Fixed #19912
- Fixed #17281
- Fixed #17429
- Fixed #17416
- Fixed #19799
- Fixed #17346
- Fixed #17968 ( multiple stuck objects are condensed into ‘multiple [object name]’

comma commented:
You sure you commited the fix for #19740? Not seeing anything about that in the diff

comma commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -62,6 +62,7 @@
 	to_chat(user, "Storage capacity: [used_capacity]/[max_capacity]GQ")
 
 // Use this proc to add file to the drive. Returns 1 on success and 0 on failure. Contains necessary sanity checks.
+var/list/badchars = list("/","\\",":","*","?","\"","<",">","|","#", ".")
```
since this list is global, it'll fail travis.
It doesn't need to be global since it's used in one place anyway, just move it inside the proc where it's used.

atlantiscze replied:
It should probably be a GLOBAL_LIST_INIT macro call
I can see this being used at more places - in the text editor programs for example, to provide feedback that file name contains forbidden characters. While this proc catches this case, it will likely produce a regular "Unable to save file" error message which may leave the user confused.

runedog48 replied:
Yes, I think the modular computers should display actual errors ( return types should be the error messages ). There have been times when using the computers when I've encountered an error but have no clue what /actually/ went wrong.

runedog48 commented:
Damnit, I can't believe I'm going to have to make another commit to fix that. And no, #19740 is not fixed, I forgot to remove it from the list. It turns out that the changes to reagents completely borked that system with the blood data and I don't know of a way to easily convert a string into an actual reagent that I can get the name from.

Looks like I'm gonna have to fix 10 more bugs to push this next commit.

BlueNexus commented:
Or just use
```
git commit --amend
git push -f
```
@runedog48

runedog48 commented:
Yeah, I learned the hard way a long time ago not to fuck with git commands. I already planned on doing more fixes so I'll just sneak it in with my next commit.

Edit: Damnit, the commit itself mentions the commit.. that's going to bug me
Edit2: Alright, fixed it without deleting everything. OCD is no longer being triggered, but I've still left a trail of ugly git history :< 
travis will still bitch 'cause I haven't moved that list but I'll get to it later.

Techhead0 commented on code/modules/virus2/dishincubator.dm:
```diff
@@ -128,12 +128,12 @@
-				foodsupply = 100
+		if (foodsupply < 100 && beaker.reagents.has_reagent(/datum/reagent/nutriment/virus_food))
+			var/food_needed = min(10, 100 - foodsupply) / 2
+			var/food_taken = min(food_needed, beaker.reagents.get_reagent_amount(/datum/reagent/nutriment/virus_food))
```
So.... doing some math... lowest of `5`, `50-(foodsupply/2)`, and the available food in the beaker? Feels a little wonky.

(Also, `min()` takes an arbitrary number of arguments.)

runedog48 replied:
Everything on the list has been tested in-game. The code seems to work just fine. 5u virus_food = 10 foodsupply.

Techhead0 replied:
(Assuming all the math is right...)
Better to drop `food_needed` and just have:
```
var/food_taken = min(5, (100 - foodsupply) / 2, beaker.reagents.get_reagent_amount(/datum/reagent/nutriment/virus_food))
```

Techhead0 commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -234,9 +234,17 @@
+				for(var/obj/embedded in embedlist)
+					if(!parsedembed.len || (!parsedembed.Find(embedded.name) && !parsedembed.Find("multiple [embedded.name]")))
+						parsedembed.Add(embedded.name)
+					else if(!parsedembed.Find("multiple [embedded.name]"))
```
You're searching for this string more than once, that's probably a sign your logic can be condensed.

runedog48 replied:
I'll look into condensing it in my next commit.

Techhead0 replied:
Yeah, this is still messy.

Techhead0 commented on code/game/objects/items/devices/chameleonproj.dm:
```diff
@@ -140,7 +140,8 @@
 			else
 				spawn(25) can_move = 1
-		step(src, direction)
+		if(!loc.density)
```
I suspect this will only keep you from moving if your *current* turf is dense.

runedog48 replied:
If the projected object is inside a wall, (closed) locker, whatever it may be, you will not be able to move. I haven't been able to think of any detrimental effects that would be caused by this.

Edit: I also tried doing it by checking if where you're going is dense or not, but you could still move out of the closet. I don't think having both checks will be necessary.

Techhead0 replied:
Detrimental effects? If you can move into the wall while cloaked, then once you're in the wall, you decloak and hop out the other side. Also... moving around normally and accidentally getting temporarily stuck in walls.

runedog48 replied:
I'm not understanding where you're coming from. If you stand in a locker and close it you can move onto any adjacent tile regardless of whether you can actually go there. The changes that I made prevent the 'dummy' hologram from moving if it's inside a dense space ( EX: Closed lockers, a wall tile, inside a disposal bin, etc. ). What I was saying is that I do not see any detrimental effects of the dummy not being able to move if it is inside a dense space ( my change ).

Techhead0 replied:
Oh, it requires a locker to see the bug. Carry on then.

Hubblenaut replied:
Why check for density instead of loc not being a turf? Seems like a risky heuristic.

Commit:
- Copied papers will no longer appear blank
- Corrected fix to #19799
- Fixed #17326
- Fixed #18008
- Fixed #19897
- Fixed #17609

runedog48 commented on code/modules/multiz/movement.dm:
```diff
@@ -191,7 +191,7 @@
 
 /atom/movable/proc/handle_fall(var/turf/landing)
-	Move(landing)
+	forceMove(landing)
```
Don't like this too much myself, the point was to bypass the 'bump' proc for the mob. The only downside is that I can find is that you can overlap when going down stairs. I'm looking into it more but for now, this will fix getting gibbed.

Techhead0 replied:
The simple change to handle both described scenarios is
```
 	if(locate(/obj/structure/stairs) in landing)
 		Move(landing)
		return 1
 	else
		forceMove(landing)
		handle_fall_effect(landing)
```

PsiOmegaDelta commented on code/game/objects/items/weapons/defib.dm:
```diff
@@ -441,18 +441,18 @@
 	audible_message("<b>\The [src]</b> [message]", "\The [src] vibrates slightly.")
 
-/obj/item/weapon/shockpaddles/emag_act(mob/user)
+/obj/item/weapon/shockpaddles/emag_act(var/uses, var/mob/user, var/obj/item/weapon/defibrillator/base)
```
You need to adhere to the standard emag_act() definition in case something calls emag on the paddles directly.
In this particular case it means you need to handle `base` being null.

comma replied:
addressed

Techhead0 commented:
Cannot merge due to conflicts. Vents/scrubbers received a more proper fix in another PR.

Commit:
Minor Changes to fixes
- Slight change to #19912 fix
- Slight change to #17281 fix
- Partial 'merge' of #19997 to resolve conflicts

runedog48 commented:
Glad lazily copy/pasting fixed that conflict. Send it through, chief.

Edit: @PsiOmegaDelta

Chiquiii commented:
Ping.

runedog48 commented:
Alright. It's been almost a week since I've updated my PR and nobody has signed off on it. I'm starting to get aggravated. It's all BUG FIXES, no game play changes or additions I don't see what the hold up is. There's been more than 30 pull requests that have been merged since I've updated my PR.

@PsiOmegaDelta 
@Techhead0 
@atlantiscze

Techhead0 commented:
I posted comments above two days ago.

comma reviewed (approved):
Won't get better with age

PsiOmegaDelta commented:
There are still unaddressed comments from Techhead0.

PurpleMartinJCK commented:
Still comments that need to be addressed

comma commented:
Those comments are about making things prettier mostly, they're not even in red. I don't think author is coming back, and I'd rather not have them block dozen of bugfixes
